Add cast functions to support 'cast as' statements

OlivierCavadenti opened the pull request:
- First draft to add knex.cast functions on Knex, like knex.raw and knex.ref but needs some work to be finished (I push to have comments, if any).
- Postgres integration test is ok, other dialects seems have issues. Need to check:
    - [x] Sqlite3
    - [x] OracleDB
    - [x] MSSQL
    - [x] Mysql
    - [x] CoachroachDB
- [x] Index.d.ts
- [ ] Documentation
- Closes https://github.com/knex/knex/issues/2587
- Closes https://github.com/knex/knex/issues/501

OlivierCavadenti committed:
knex.cast function to support 'cast as' statements

OlivierCavadenti committed:
clean up

kibertoad commented on lib/cast.js:
```diff
@@ -0,0 +1,132 @@
+  _getColType(type) {
+    let columnCompiler;
+
+    switch (this.client.config.client) {
```
why not implement getCompiler method on this.client instead? not sure we don't have one yet, actually

kibertoad commented on lib/client.js:
```diff
@@ -141,6 +143,47 @@ class Client extends EventEmitter {
   }
+
+  cast(value, to, alias) {
+    return new Cast(this).cast(value, to, alias);
```
if new Cast is only supposed to be called from client, then you can actually pass correct compiler from here as a constructor param

OlivierCavadenti replied:
I try quickly but it's seems I cant use columnBuilder function because it's depends of args from tableBuilder, type, args. But maybe you want getCompiler function (overrided in each Client) that return the columnCompiler prototype ?

OlivierCavadenti committed:
refacto getColType + merge master

OlivierCavadenti committed:
remove unused require

OlivierCavadenti committed:
fix tests excepted MSSQL

OlivierCavadenti committed:
to const

OlivierCavadenti commented:
I have strange behavior with MSSQL when I cast binary:

![image](https://user-images.githubusercontent.com/10152806/158063734-c8832468-a0be-41aa-ad1d-203fa88dd4fe.png)

Can't explain that, if anyone have clues...

OlivierCavadenti committed:
tu fixes

OlivierCavadenti committed:
remove only

OlivierCavadenti committed:
index.d.ts for cast functions

OlivierCavadenti committed:
fix: remove leftover conflict marker in mysql/index.js

OlivierCavadenti committed:
fix: add OracleDB and MSSQL cast support, fix castBigint and castBinary

- Fix _getColType fallback for 'bigint' → 'biginteger' (Oracle, SQLite)
- Add castConvert for MSSQL binary to avoid UTF-16LE encoding issue
- Add OracleDB overrides for castJson/castJsonb (→ clob)
- Add MSSQL override for castBinary (→ convert instead of cast)
- Add unit tests for all cast functions across all dialects
- Add OracleDB integration test for cast functions
- Fix MSSQL integration test expected SQL for castBinary

OlivierCavadenti committed:
Merge branch 'master' into cast-support

tgriesser committed:
Merge branch 'master' into cast-support

OlivierCavadenti committed:
docs: document cast functions in the query builder guide

Add reference entries for castInt/castBigint/castDouble/castReal/castText/castChar/castJson/castJsonb/castBinary, matching the existing json* documentation style, with the MySQL castText limitation and the MSSQL binary double-cast noted.

tgriesser commented on lib/cast.js:
```diff
@@ -0,0 +1,116 @@
+
+// Allow the `Cast` object to be utilized with full access to the relevant
+// promise API.
+augmentWithBuilderInterface(Cast);
```
Why does `Cast` require the full promise API?

OlivierCavadenti replied:
Good catch, it doesn't. `Cast` extends `Raw`, and `Raw` already calls `augmentWithBuilderInterface` / `addQueryContext` and sets `isRawInstance`, so re-applying them here was pure redundancy. I removed them.

tgriesser commented on lib/cast.js:
```diff
@@ -0,0 +1,116 @@
+  _getColType(type) {
+    const columnCompiler = this.client.getColumnCompiler();
+    const typeName =
+      columnCompiler.prototype[type] ?? columnCompiler.prototype[type + 'eger'];
```
```suggestion
columnCompiler.prototype[type + 'eger'];
```
Feels kinda odd, I didn't see anywhere that this is doing `_getColType('int')` which is what I presume this is attempting to do?

OlivierCavadenti replied:
You're right that it's confusing ! fixed

tgriesser reviewed (commented):
Nice idea, a few comments

OlivierCavadenti committed:
refactor(cast): drop redundant Raw augmentation and simplify _getColType

Cast extends Raw, which already applies augmentWithBuilderInterface,
addQueryContext and sets isRawInstance, so re-applying them on Cast was
redundant.

_getColType no longer falls back to `type + 'eger'`; that hack only
existed because castBigint passed 'bigint' instead of the actual column
compiler prototype key 'biginteger'. Use the real key and drop the
fallback.

tgriesser committed:
Merge branch 'master' into cast-support

OlivierCavadenti committed:
fix(mariadb): emit MariaDB-valid SQL for cast helpers

MariaDB does not support `CAST(... AS JSON)` (its JSON type is an alias
for LONGTEXT) nor `CAST(... AS REAL)` (REAL is a synonym for DOUBLE),
so the castJson/castJsonb/castReal helpers produced statements that
MariaDB rejected with a syntax error, breaking the new mariadb CI matrix.

Override these three helpers in the MariaDB client to cast to `char`
(json/jsonb) and `double` (real), which MariaDB accepts. The other cast
targets (double, char, unsigned, binary) are already valid on MariaDB.

tgriesser commented on lib/dialects/mysql/index.js:
```diff
@@ -344,6 +349,32 @@ class Client_MySQL extends Client {
+  castText(value, alias) {
+    throw new Error(
+      `Cast to Test is not possible in MySQL, use castChar function instead.`
+    );
```
Should we just have this `castChar` - would seem more consistent with knex's behavior elsewhere, to sort of smooth over cross-dialect inconsistencies

OlivierCavadenti replied:
Agreed, that's more consistent. castText now produces a CHAR cast on MySQL instead of throwing, so it behaves like the other dialects.

tgriesser commented on types/index.d.ts:
```diff
@@ -3365,6 +3376,7 @@ declare namespace Knex {
     runner(builder: any): any;
     transaction(container: any, config: any, outerTx: any): Transaction;
     raw(...args: any[]): any;
+    castText(...args: any[]): any;
```
```suggestion
```

Do we want to put this on the `client` interface? If so should we add all of the `cast*` fns?

OlivierCavadenti replied:
Removed it. The cast helpers are already typed on the public interface around line 2429, so castText alone on the internal Client type was redundant.

OlivierCavadenti committed:
fix(mysql): map castText to CHAR instead of throwing

Smooths the cross-dialect inconsistency so castText works on MySQL
like the other dialects. Also drop the redundant castText from the
internal Client type (already typed on the public interface).

tgriesser commented on test/unit/query/builder.js:
```diff
@@ -12208,4 +12208,125 @@ describe('QueryBuilder', () => {
+        pg: 'select cast(? as integer) from "t"',
+        mysql: 'select cast(? as unsigned) from `t`',
+        mssql: 'select cast(? as int) from [t]',
+      });
```
Can we add `mariadb` to the tests here?

OlivierCavadenti replied:
Added mariadb to the cast tests. It matches mysql except for castJson and castJsonb, which map to char, and castReal, which maps to double, reflecting the mariadb overrides.

OlivierCavadenti committed:
Merge remote-tracking branch 'upstream/master' into cast-support

OlivierCavadenti committed:
test: cover mariadb in cast unit tests
